File: 76-minimum-window-substring/minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        minLength = float("infinity")
        res = ""

        tChars = {}
        for char in t:
            tChars[char] = tChars.get(char, 0) + 1
        
        need = len(tChars.keys())
        have = 0

        start = 0
        window = {}

        for end, char in enumerate(s):
            window[char] = window.get(char, 0) + 1
            if char in tChars and window[char] == tChars[char]:
                have += 1

            while have >= need:
                minLength = min(minLength, end - start + 1)
                if minLength == (end - start + 1): 
                    res = s[start: end + 1]

                startChar = s[start]
                window[startChar] -= 1

                if startChar in tChars and window[startChar] < tChars[startChar]:
                    have -= 1
                    
                start += 1
        
        return res

File: 76-minimum-window-substring/minimum-window-substring.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {
  if (t === "") return "";

  let countT = {};
  let window = {};

  for (let i = 0; i < t.length; i++){
      const char = t[i];
      if (char in countT) {
          countT[char]++; 
      } else {
          countT[char] = 1
      }
  }

  let have = 0;
  const need = Object.keys(countT).length; 

  let res = "";
  let resLength = Infinity; 

  let left = 0;

  for (let right = 0; right < s.length; right++){
      char = s[right];
      if(char in window) {
          window[char]++;
      } else {
          window[char] = 1;
      }

      if (countT[char] && (window[char] == countT[char])) {
          have += 1;
      }

      while (have >= need) {
          if ((right - left + 1) < resLength) {
              res = s.slice(left, right + 1);
              resLength = (right - left + 1)
          }

          const leftChar = s[left];
          window[leftChar] -= 1; 

          if(leftChar in countT && (window[leftChar] < countT[leftChar])) {
              have -= 1;
          }
          
          left += 1;
      } 
  }
  
    return res

};
